client/checkout/classic/test/upe-split.test.js

```diff
@@ -1,74 +1,191 @@
+			document.body.removeChild( container );
+			container = null;
+		} );
 
-	test( 'saved CC is selected', () => {
-		const input = document.querySelector(
-			'#wc-woocommerce_payments-payment-token-new'
-		);
-		input.checked = false;
-		const paymentMethodType = 'card';
+		test( 'new CC is selected', () => {
+			const input = document.querySelector(
+				'#wc-woocommerce_payments-payment-token-new'
+			);
+			input.checked = true;
+			const paymentMethodType = 'card';
 
-		expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe( true );
-	} );
+			expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe(
+				false
+			);
+		} );
+
+		test( 'saved CC is selected', () => {
+			const input = document.querySelector(
+				'#wc-woocommerce_payments-payment-token-new'
+			);
+			input.checked = false;
+			const paymentMethodType = 'card';
+
+			expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe(
+				true
+			);
+		} );
 
-	test( 'new SEPA is selected', () => {
-		const input = document.querySelector(
-			'#wc-woocommerce_payments_sepa_debit-payment-token-new'
-		);
-		input.checked = true;
-		const paymentMethodType = 'sepa_debit';
+		test( 'new SEPA is selected', () => {
+			const input = document.querySelector(
+				'#wc-woocommerce_payments_sepa_debit-payment-token-new'
+			);
+			input.checked = true;
+			const paymentMethodType = 'sepa_debit';
 
-		expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe( false );
+			expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe(
+				false
+			);
+		} );
+
+		test( 'saved SEPA is selected', () => {
+			const input = document.querySelector(
+				'#wc-woocommerce_payments_sepa_debit-payment-token-new'
+			);
+			input.checked = false;
+			const paymentMethodType = 'sepa_debit';
+
+			expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe(
+				true
+			);
+		} );
+
+		test( 'non-tokenized payment gateway is selected', () => {
+			const paymentMethodType = 'sofort';
+
+			expect( isUsingSavedPaymentMethod( paymentMethodType ) ).toBe(
+				false
+			);
+		} );
 	} );
 
-	test( 'saved SEPA is selected', () => {
-		const input = document.querySelector(
-			'#wc-woocommerce_payments_sepa_debit-payment-token-new'
-		);
-		input.checked = false;
-		const paymentMethodType = 'sepa_debit';
+	describe( 'getSelectedUPEGatewayPaymentMethod', () => {
+		let container;
+		let input;
+
+		beforeAll( () => {
+			container = document.createElement( 'div' );
+			container.innerHTML = `
+			<ul class="wc_payment_methods payment_methods methods">
+				<li class="wc_payment_method payment_method_woocommerce_payments">
+					<input id="payment_method_woocommerce_payments" type="radio" class="input-radio">
+				</li>
+				<li class="wc_payment_method payment_method_woocommerce_payments_bancontact">
+					<input id="payment_method_woocommerce_payments_bancontact" type="radio" class="input-radio">
+				</li>
+			</ul>
+			`;
+			document.body.appendChild( container );
+		} );
```

Comment:
It's a small suggestion, but couldn't you move the `beforeAll` and `afterAll` from both child `describe` blocks (`isUsingSavedPaymentMethod` and `getSelectedUPEGatewayPaymentMethod`) and change them to `beforeEach` and `afterEach` respectively? That way, we would create one container for each `describe` test block, remove it upon completion of those particular tests, and then recreate it for the next block.

```js
describe( 'UPE split checkout', () => {
	let container;
	
	beforeEach( () => {
		container = document.createElement( 'div' );
		container.innerHTML = `
			<ul class="wc_payment_methods payment_methods methods">
				<li class="wc_payment_method payment_method_woocommerce_payments">
					<input id="payment_method_woocommerce_payments" type="radio" class="input-radio">
				</li>
				<li class="wc_payment_method payment_method_woocommerce_payments_bancontact">
					<input id="payment_method_woocommerce_payments_bancontact" type="radio" class="input-radio">
				</li>
			</ul>
		`;
		document.body.appendChild( container );
	} );

	afterEach( () => {
		document.body.removeChild( container );
		container = null;
	} );
	
	describe( 'isUsingSavedPaymentMethod', () => {
		// isUsingSavedPaymentMethod tests...
	}
	
	describe( 'getSelectedUPEGatewayPaymentMethod', () => {
		// getSelectedUPEGatewayPaymentMethod tests...
	}
	
	describe( 'getSetupIntentFromSession', () => {
		// getSelectedUPEGatewayPaymentMethod tests...
	}
}
```

You don't need the container for the final group of tests (`getSetupIntentFromSession`), but I suppose it doesn't hurt. Otherwise if you want to avoid creating this container for the third block of tests, I would extract the `beforeAll` and `afterAll` functionality into a separate function, so that you could reuse it in the first two blocks of tests.

```js
let container;
const setUpContainer = () => {
	container = document.createElement( 'div' );
	container.innerHTML = `
		<ul class="wc_payment_methods payment_methods methods">
			<li class="wc_payment_method payment_method_woocommerce_payments">
				<input id="payment_method_woocommerce_payments" type="radio" class="input-radio">
			</li>
			<li class="wc_payment_method payment_method_woocommerce_payments_bancontact">
				<input id="payment_method_woocommerce_payments_bancontact" type="radio" class="input-radio">
			</li>
		</ul>
	`;
	document.body.appendChild( container );
}
const tearDownContainer = () => {
	document.body.removeChild( container );
	container = null;
}
describe( 'UPE split checkout', () => {

	describe( 'isUsingSavedPaymentMethod', () => {
		beforeAll( setUpContainer );
		afterAll( tearDownContainer );
		// isUsingSavedPaymentMethod tests...
	}
	
	describe( 'getSelectedUPEGatewayPaymentMethod', () => {
		beforeAll( setUpContainer );
		afterAll( tearDownContainer );
		// getSelectedUPEGatewayPaymentMethod tests...
	}
	
	describe( 'getSetupIntentFromSession', () => {
		// getSelectedUPEGatewayPaymentMethod tests...
	}
}
```

This basically allows us to avoid having copied code in the same file. Let me know your thoughts and if either of these approaches takes your fancy. Cheers!

Reply:
@FangedParakeet I think this would make sense if the both blocks of mocked HTML were the same. One block is mocking the new payment method radio button with the "Use a new payment method" label,  while the other block is mocking the radio button for individual payment method selection. Is the suggestion here to combine the 2 HTML blocks? If so, that could work.

Reply:
Ah, I see. This is another classic case of me not properly reading something right in front of me. I thought those HTML container elements were basically the same, but in fact they're drastically different. In that case, continue on as normal, nothing to see here. My bad! 😅